diffrax/_solver/srk.py

```diff
@@ -0,0 +1,624 @@
+                # _wgs = Array[wg1, wg2, ..., wg{j-1}, 0, 0, ..., 0]
+                # _levy_gs_list = [H_gs, K_gs] (if those entries exist) where
+                # H_gs = Array[Hg1, Hg2, ..., Hg{j-1}, 0, 0, ..., 0]
+                # K_gs = Array[Kg1, Kg2, ..., Kg{j-1}, 0, 0, ..., 0]
+                # tfi = drift.vf_prod(t0 + c_i*h, y_i, args, h) (i.e. hki = h * k_i)
+                # wgi = diffusion.vf_prod(t0 + c_i*h, y_i, args, w) (i.e. wgi = w * g_i)
+
+                wg_sum = sum_prev_stages(_wgs, a_w_j)
+                levy_sum_list = [
+                    sum_prev_stages(levy_gs, a_levy_j)
+                    for a_levy_j, levy_gs in zip(a_levy_list_j, _levy_gs_list)
+                ]
+
+                _diffusion_result = jtu.tree_map(
+                    lambda _wg, *_levy_g: sum(_levy_g, _wg), wg_sum, *levy_sum_list
+                )
+
+            # compute Σ_{i=1}^{j-1} a_j_i tf_i
+            _drift_result = sum_prev_stages(_tfs, a_j)
+
+            # z_j = y_0 + h (Σ_{i=1}^{j-1} a_j_i k_i) + g * (a_w_j * ΔW + cH_j * ΔH)
+            z_j = (y0**ω + _drift_result**ω + _diffusion_result**ω).ω
+
+            def insert_jth_stage(results, res_j):
+                return jtu.tree_map(
+                    lambda res_leaf, j_leaf: res_leaf.at[j].set(j_leaf), results, res_j
+                )
+
+            def compute_and_insert_tfj(_tfs_in):
+                tf_j = drift.vf_prod(t0 + c_j * h, z_j, args, h)
+                return insert_jth_stage(_tfs_in, tf_j)
+
+            if self.tableau.ignore_stage_f is None:
+                _tfs = compute_and_insert_tfj(_tfs)
+            else:
+                drift_pred = jnp.logical_not(ignore_stage_f[j])
+                _tfs = lax.cond(drift_pred, compute_and_insert_tfj, lambda x: x, _tfs)
+
+            if additive_noise:
+                return (j + 1, _tfs, None, None), None
+
+            def compute_and_insert_gj(_wgs_in, _levy_gs_list_in):
+                wg_j = diffusion.vf_prod(t0 + c_j * h, z_j, args, w)
+                new_wgs = insert_jth_stage(_wgs_in, wg_j)
+
+                levy_g_list_j = [
+                    diffusion.vf_prod(t0 + c_j * h, z_j, args, levy)
+                    for levy in levy_areas
+                ]
+                new_levy_gs_list = insert_jth_stage(_levy_gs_list_in, levy_g_list_j)
+                return new_wgs, new_levy_gs_list
+
+            if self.tableau.ignore_stage_g is None:
+                _wgs, _levy_gs_list = compute_and_insert_gj(_wgs, _levy_gs_list)
+            else:
+                diffusion_pred = jnp.logical_not(ignore_stage_g[j])
+                _wgs, _levy_gs_list = lax.cond(
+                    diffusion_pred,
+                    compute_and_insert_gj,
+                    lambda x, y: (x, y),
+                    _wgs,
+                    _levy_gs_list,
+                )
+
+            return (j + 1, _tfs, _wgs, _levy_gs_list), None
+
+        scan_out = lax.scan(stage, carry, scan_inputs, length=len(b_sol))
```

Comment:
Note that due to an annoying long-standing JAX performance issue, when performing reverse-mode autodifferentiation on this, the `.at[].set()` you have earlier in the stage function will probably actually make an out-of-place copy, rather than actually modifying the array in-place.

This can be fixed by using `eqxi.scan(..., buffers=...)` instead. (And you'll probably want to use `checkpoints="all"` as well.)

Reply:
I tried doing that, but there is an issue. I have to dot-multiply `_a_j` with the previously computed stages. Previously, the unfilled part of _tfs and _wgs were zeros and i just did a tensordot. But with buffers I am not able to read from unfilled parts of the buffer. I also don't think I can slice arrays with the stage number because then shapes are not always the same.  I tried to make the stage number static somehow, but I'm still missing something. I think I'll need a better explanation. Could you send an example or maybe we could have a call?

Reply:
Take a look at how `AbstractRungeKutta` tackles this:

https://github.com/patrick-kidger/diffrax/blob/e33c17eeb7ca5ec081f763f6f3d146c2e8df4231/diffrax/_solver/runge_kutta.py#L861

We explicitly unwrap the whole buffer with a `[...]`. Mathematically, this works because the unfilled parts of the buffer are all filled with zero. Autodifferentiably, this works because whilst the rule with buffers is that "we musn't read from anything we haven't written to yet", this holds in the mathematical sense not the computational sense. And as above, reading those zeros doesn't actually change the result of our computation.

Indeed there'll be no way to make the stage number static.

Reply:
Took me a while to figure out that the key is to factor out the PyTree structure of ks so that `jtu.tree_map(lambda _, _buff: _buff[...], y0, _stage_out_buff)` treats the buffer as a leaf. Just looking at your comments in `runge-kutta.py`, it wasn't entirely clear to me that y_map, f_map and t_map are there precisely for factoring out the PyTree structure of y, f and the tableaus resp. Phew, happy I managed to decipher the whole idea in the end. But I suppose I learned more this way than if you gave it to me on a platter, haha.

Reply:
Haha, good job! That's exactly the right way to tackle these.